src/sokoban.cpp

```diff
@@ -27,14 +37,128 @@ std::vector<std::string> Sokoban::board() {
 }
 
 void Sokoban::locate_player() {
-    for (int y = 0; y < _board.size(); y++) {
-        for (int x = 0; x < _board[y].size(); x++) {
-            if (_board[y][x] == '@' || _board[y][x] == '+') {
+    for (unsigned int y = 0; y < _board.size(); y++) {
+        for (unsigned int x = 0; x < _board[y].size(); x++) {
+            if (_board[y][x] == Cell::PLAYER || 
+                _board[y][x] == Cell::PLAYER_ON_GOAL) {
                 px = x;
                 py = y;
                 return;
             }
         }
     }
     throw std::invalid_argument("Player not found...");
+}
+
+void Sokoban::move_player(int dy, int dx) {
+    char current_cell = (_board[py][px] == Cell::PLAYER_ON_GOAL) ? 
+        Cell::GOAL : Cell::EMPTY;
+    char next_cell = (_board[py+dy][px+dx] == Cell::GOAL) ? 
+        Cell::PLAYER_ON_GOAL : Cell::PLAYER;
+
+    _board[py][px] = current_cell;
+    _board[py+dy][px+dx] = next_cell;
+    py += dy;
+    px += dx;
+}
+
+void Sokoban::push_box(int dy, int dx) {
+    // Set the box's new and old cells
+    _board[py+dy][px+dx] = (_board[py+dy][px+dx] == Cell::BOX_ON_GOAL) ? 
+        Cell::GOAL : Cell::EMPTY;
+    _board[py+dy+dy][px+dx+dx] = (_board[py+dy+dy][px+dx+dx] == Cell::GOAL) ? 
+        Cell::BOX_ON_GOAL : Cell::BOX;
+}
+
+void Sokoban::pull_box(int dy, int dx) {
+    // Set the box's new and old cells
+    _board[py-dy-dy][px-dx-dx] = (_board[py-dy-dy][px-dx-dx] == Cell::BOX_ON_GOAL) ? 
+        Cell::GOAL : Cell::EMPTY;
+    _board[py-dy][px-dx] = (_board[py-dy][px-dx] == Cell::GOAL) ? 
+        Cell::BOX_ON_GOAL : Cell::BOX;
+}
+
+bool Sokoban::make_move(Direction direction) {
+    std::vector<int> offsets = dir_offset[direction];
```

Comment:
I think `int [dy, dx] = dir_offset[direction];` should work here.

Reply:
Apparently it doesn't, but `pair` does:

```cpp
        std::unordered_map<Direction, std::pair<int, int>> directions = {
            {U, std::make_pair(-1, 0)},
            {D, std::make_pair(1, 0)},
            {L, std::make_pair(0, -1)},
            {R, std::make_pair(0, 1)},
        };
        auto [dy, dx] = directions.at(dir);
```

I'm fine leaving it with the original though, the map is the most important part.

Reply:
It didn't work for vector. But I think it would work for pair, tuple, or struct.